libbpf-tools/biostacks.c

```diff
@@ -0,0 +1,213 @@
+{
+}
+
+static int print_map(struct ksyms *ksyms, int fd)
+{
+	char *units = env.milliseconds ? "msecs" : "usecs";
+	struct rqinfo lookup_key = {}, next_key;
+	const struct ksym *ksym;
+	int num_stack, i, err;
+	struct hist hist;
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		err = bpf_map_lookup_elem(fd, &next_key, &hist);
+		if (err < 0) {
+			fprintf(stderr, "failed to lookup hist: %d\n", err);
+			return -1;
+		}
+		printf("%-14.14s %-6d %-7s\n",
+			next_key.comm, next_key.pid, next_key.disk);
+		num_stack = next_key.kern_stack_size /
+			sizeof(next_key.kern_stack[0]);
+		for (i = 0; i < num_stack; i++) {
+			ksym = ksyms__map_addr(ksyms, next_key.kern_stack[i]);
+			printf("%s\n", ksym->name);
+		}
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		printf("\n");
+		lookup_key = next_key;
+	}
+
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	static const struct argp argp = {
+		.options = opts,
+		.parser = parse_arg,
+		.doc = argp_program_doc,
+	};
+	struct ksyms *ksyms = NULL;
+	struct biostacks_bpf *obj;
+	int err;
+
+	err = argp_parse(&argp, argc, argv, 0, NULL, NULL);
+	if (err)
+		return err;
+
+	libbpf_set_print(libbpf_print_fn);
+
+	err = bump_memlock_rlimit();
+	if (err) {
+		fprintf(stderr, "failed to increase rlimit: %d\n", err);
+		return 1;
+	}
+
+	obj = biostacks_bpf__open();
+	if (!obj) {
+		fprintf(stderr, "failed to open and/or load BPF ojbect\n");
+		return 1;
+	}
+
+	/* initialize global data (filtering options) */
+	if (env.disk)
+		strncpy((char*)obj->rodata->targ_disk, env.disk, env.disk_len);
+	obj->rodata->targ_ms = env.milliseconds;
+
+	err = biostacks_bpf__load(obj);
+	if (err) {
+		fprintf(stderr, "failed to load BPF object: %d\n", err);
+		goto cleanup;
+	}
+
+	obj->links.fentry__blk_account_io_start =
+		bpf_program__attach(obj->progs.fentry__blk_account_io_start);
+	err = libbpf_get_error(obj->links.fentry__blk_account_io_start);
+	if (err) {
+		fprintf(stderr, "failed to attach blk_account_io_start: %s\n",
+			strerror(err));
+		goto cleanup;
+	}
+	ksyms = ksyms__load();
+	if (!ksyms) {
+		fprintf(stderr, "failed to load kallsyms\n");
+		goto cleanup;
+	}
+	if (ksyms__get_symbol(ksyms, "blk_account_io_merge_bio")) {
+		obj->links.kprobe__blk_account_io_merge_bio =
+			bpf_program__attach(obj->
+					progs.kprobe__blk_account_io_merge_bio);
+		err = libbpf_get_error(obj->
+				links.kprobe__blk_account_io_merge_bio);
+		if (err) {
+			fprintf(stderr, "failed to attach "
+				"blk_account_io_merge_bio: %s\n",
+				strerror(err));
+			goto cleanup;
+		}
+	}
+	obj->links.fentry__blk_account_io_done =
+		bpf_program__attach(obj->progs.fentry__blk_account_io_done);
+	err = libbpf_get_error(obj->links.fentry__blk_account_io_done);
+	if (err) {
+		fprintf(stderr, "failed to attach blk_account_io_done: %s\n",
+			strerror(err));
+		goto cleanup;
+	}
+
+	signal(SIGINT, sig_handler);
+
+	printf("Tracing block I/O with init stacks. Hit Ctrl-C to end.\n");
+	sleep(env.duration);
```

Comment:
this will return prematurely if Ctrl-C hits, is that the intent?

Reply:
Yes, if the user does not specify a duration,  it will sleep forever, so use ctrl-c to end the sleep and output the result.